File: java/goodmem-semantic-kernel/src/main/java/ai/goodmem/semantickernel/GoodMemUpsertException.java
package ai.goodmem.semantickernel;

/**
 * Thrown when an upsert fails while replacing an existing record.
 *
 * <p>GoodMem has no update endpoint ({@code PUT} and {@code PATCH} return 404) and
 * rejects a create that reuses an existing ID with 409, so replacing a record means
 * deleting the old memory before writing the new one. When the write fails, the
 * connector puts the previous version back; {@link #isRestored()} says whether that
 * succeeded and {@link #getLostKey()} names the record if it did not.
 */
public class GoodMemUpsertException extends GoodMemException {

    private final boolean restored;
    private final String lostKey;

    public GoodMemUpsertException(String message, boolean restored, String lostKey, Throwable cause) {
        super(message, cause);
        this.restored = restored;
        this.lostKey = lostKey;
    }

    /** Whether the previous version was successfully restored. */
    public boolean isRestored() {
        return restored;
    }

    /** The key whose previous version could not be restored, or {@code null}. */
    public String getLostKey() {
        return lostKey;
    }
}

File: dotnet/GoodMem.SemanticKernel/GoodMemUpsertException.cs
using System;

namespace GoodMem.SemanticKernel;

/// <summary>
/// Thrown when an upsert fails while replacing an existing record.
/// </summary>
/// <remarks>
/// GoodMem has no update endpoint (<c>PUT</c> and <c>PATCH</c> return 404) and
/// rejects a create that reuses an existing id with 409, so replacing a record
/// means deleting the old memory before writing the new one. When the write
/// fails, the connector puts the previous version back; <see cref="Restored"/>
/// says whether that succeeded and <see cref="LostKey"/> names the record if it
/// did not.
/// </remarks>
public sealed class GoodMemUpsertException : Exception
{
    /// <summary>Whether the previous version was successfully restored.</summary>
    public bool Restored { get; }

    /// <summary>The key whose previous version could not be restored, if any.</summary>
    public string? LostKey { get; }

    /// <summary>Initialises the exception.</summary>
    public GoodMemUpsertException(string message, bool restored, string? lostKey, Exception? inner = null)
        : base(message, inner)
    {
        Restored = restored;
        LostKey = lostKey;
    }
}
